sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,190 @@
+                             "of bins. Received {0}, expected at least 2."
+                             .format(self.n_bins))
+
+        X = check_array(X, accept_sparse=self.sparse_formats, dtype=float,
+                        ensure_2d=False)
+        self.n_features_ = X.shape[1]
+
+        self.min_ = X.min(axis=0)
+        self.max_ = X.max(axis=0)
+        return self
+
+
+    def transform(self, X, y=None):
+        """Discretizes the input data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse-matrix}, shape [n_samples, n_features]
+            The input data containing continuous features to be
+            discretized. Categorical columns are indicated by
+            categorical_columns from the constructor.
+
+        Returns
+        -------
+        T : array-like, shape [n_samples, n_features]
+            Array containing discretized features alongside the
+            categorical features. Categorical features are represented
+            as the last features of the output.
+
+        """
+        check_is_fitted(self, ["min_", "max_"])
+        X = check_array(X, accept_sparse=self.sparse_formats, dtype=float,
+                        ensure_2d=False)
+
+        if X.shape[1] != self.n_features_:
+            raise ValueError("Transformed array does not have currect number "
+                             "of features. Expecting {0}, received {1}."
+                             .format(self.n_features_, X.shape[1]))
+
+        # TODO: Make copy parameter?
+        X = X.copy()
+
+        # Clip outliers from data (TODO)
+        #X[X > self.max_] = self.max_
+        #X[X < self.min_] = self.min_
+
+        return self._transform(X)
+
+
+    def _transform(self, X):
+        """Discretizes the continuous features into k integers.
+
+        The process used below preserves the sparsity of the data by
+        transforming the data, and rounding the data down to the
+        nearest integer. These integers represent the bins sought.
+
+        The process is as follows.
+        1. Let X_trans = X * n_bins / (X.max() - X.min()).
+            Observe that X_trans is a dataset of width k.
+        2. epsilon = abs(X_trans.min() - ceil(X_trans.min())).
+            episilon < 1 is the distance of X.min() to ceil(X.min())
+        3. X_trans[X_trans != 0] += epsilon
+            Shifts the data. X_trans.min() is now an integer. Note this
+            operation can be done on sparse matrices.
+        4. output = floor(X_trans)
+            The output now contains integers which represent the
+            discretized values. Observe that even if we added epsilon to
+            zero values in step 3, these zeros will still be discretized
+            to zero.
+        """
+        # Rescale data
+        X_trans = X / (self.max_ - self.min_)
+        X_trans *= self.n_bins
+
+        # Shift data
+        epsilon = np.abs(X_trans.min() - np.ceil(X_trans.min()))
+        X_trans[X_trans != 0] += epsilon
```

Comment:
@amueller You mentioned that you wanted this discretizer to mask out categorical features. My previous implementation in #5825 moved categorical features to the last columns of `X`. (Example, if there were 3 categorical features, after the transformation, they would appear as the last 3 columns in `X_trans`.) I'm trying to avoid that in this implementation. 

But the above lines in my code are built to work with sparse matrices. I'm thinking the following code to have it implemented by specifying which continuous features to bin using the code below:

```
from scipy.sparse import csr_matrix
import numpy as np

matrix = csr_matrix([[0, 1, 2, 3],
                     [0, 0, 4, 0],
                     [5, 0, 0, 0]], dtype=float)
continuous = [0, 3]
epsilon = 0.5

submatrix = matrix[:, continuous]  # Get only the continuous features
submatrix[submatrix != 0] += epsilon  # The shift step

matrix[:, continuous] = submatrix
```

which looks reasonable. Except scipy 0.11.0 raises a `NotImplementedError` on this line:

```
submatrix[submatrix != 0] += epsilon  # The shift step
```

Do you have work arounds for "masking out" categorical features? (And allowing manipulation on only continuous features.)